Formal modeling of view change + checkpointing

HristoStaykov opened the pull request:
* **Problem Overview**  
  After adding the Checkpoint generation in the model we need to reflect it in the View Change process.
  The work is still in progress.
* **Testing Done**  
 NA

teoparvanov commented on docs/sbft-formal-model/messages.dfy:
```diff
@@ -73,13 +87,21 @@ module Messages {
+                         && (NewViewMsg? ==> vcMsgs.valid(newView, quorumSize))
                        }
                      }
+  predicate CheckMessageValidity(msg:Message, quorumSize:nat) {
```
Maybe `IsMessageValid` is a better name for a predicate?

HristoStaykov replied:
This is a temporary fix resolving a problem with Dafny's recursion expansion heuristics. Yesterday we had discussion with Oded Padon and he suggested a much better solution which I have added in the latest commit.

teoparvanov replied:
Thanks for the clarification!

teoparvanov reviewed (commented):
Thanks @HristoStaykov, this is a very elegant piece of formal modeling! I see some TODOs in the code, but are the invariant proofs successful at this point? You can mention it in the "Testing done" section. 

So, what are the remaining steps as part of this PR?

HristoStaykov commented:
> Thanks @HristoStaykov, this is a very elegant piece of formal modeling! I see some TODOs in the code, but are the invariant proofs successful at this point? You can mention it in the "Testing done" section.
> 
> In summary, what are the remaining steps as part of this PR?

After we have added the necessary changes to reflect the Checkpointing into the View Change process we need to debug the current state of the proof and resolve 2 problems related to the LeaveView step, in which we need to clear the contents of the Working Window (we have un-commented the steps related to View Change). This is problematic with the following 2 predicates we have in the Invariant - **SentPreparesMatchRecordedPrePrepareIfHostInSameView** and **EveryCommitIsSupportedByRecordedPrepares**

teoparvanov commented:
> > Thanks @HristoStaykov, this is a very elegant piece of formal modeling! I see some TODOs in the code, but are the invariant proofs successful at this point? You can mention it in the "Testing done" section.
> > In summary, what are the remaining steps as part of this PR?
> 
> After we have added the necessary changes to reflect the Checkpointing into the View Change process we need to debug the current state of the proof and resolve 2 problems related to the LeaveView step, in which we need to clear the contents of the Working Window (we have un-commented the steps related to View Change). This is problematic with the following 2 predicates we have in the Invariant - **SentPreparesMatchRecordedPrePrepareIfHostInSameView** and **EveryCommitIsSupportedByRecordedPrepares**

Thanks Hristo, just wondering - is it possible that these two intermediate invariants would need to be formulated differently, with view change in the picture? Maybe some of the previous claims don't hold anymore?

teoparvanov commented on docs/sbft-formal-model/messages.dfy:
```diff
@@ -58,27 +56,63 @@ module Messages {
+      && (forall m | m in msgs :: && m.payload.CheckpointMsg?
+                                  && m.payload == prototype()
+                                  && m.payload.seqIDReached == lastStableCheckpoint)
+      && UniqueSenders(msgs)
```
It'd be cleaner to move this condition first, right before `&& |msgs| > 0`

teoparvanov commented on docs/sbft-formal-model/proof.dfy:
```diff
@@ -765,9 +773,155 @@ module Proof {
+    requires SendNewViewMsgStepIsEnabled(c, v, v', step, h_step)
+    ensures Inv(c, v')
+  {
+    CommitMsgStability(c, v, v', step);
```
Shouldn't we also model view change-specific invariants, like the one stating that a message committed in the previous view is also committed in the new view?

HristoStaykov replied:
Yes, definitely this will be our next target invariant. But before we introduce it we need to fix all the existing ones as described above.

teoparvanov commented on docs/sbft-formal-model/replica.i.dfy:
```diff
@@ -42,59 +42,80 @@ module Replica {
+    map[]
+  }
+
+  // Define your Host protocol state machine here.
```
nit - we can probably remove this comment?

HristoStaykov replied:
I will restructure the comments according to the context.

teoparvanov commented on docs/sbft-formal-model/replica.i.dfy:
```diff
@@ -42,59 +42,80 @@ module Replica {
+  datatype Constants = Constants(myId:HostId, clusterConfig:ClusterConfig.Constants) {
+    // host constants coupled to DistributedSystem Constants:
+    // DistributedSystem tells us our id so we can recognize inbound messages.
+    // clusterSize is in clusterConfig.
```
Same here, the comment doesn't seem to match the code very well...

HristoStaykov replied:
As above I will introduce clarifications and move the comments to the intended context.

teoparvanov commented on docs/sbft-formal-model/replica.i.dfy:
```diff
@@ -360,10 +434,9 @@ module Replica {
+                                                                          msg.payload.operationWrapper])
   }
 
   predicate ContiguousCommits(c:Constants, v:Variables, targetSeqID:SequenceID)
```
For better readability I'd call this `ContiguousCommitsUntil`. WDYT?

HristoStaykov replied:
Yes, we need to to be more concise, thanks.

teoparvanov reviewed (commented):
Great progress @HristoStaykov ! Is the state transfer step (currently TODO) - also needed, so that the invariant proofs succeed?

HristoStaykov commented:
> > > Thanks @HristoStaykov, this is a very elegant piece of formal modeling! I see some TODOs in the code, but are the invariant proofs successful at this point? You can mention it in the "Testing done" section.
> > > In summary, what are the remaining steps as part of this PR?
> > 
> > 
> > After we have added the necessary changes to reflect the Checkpointing into the View Change process we need to debug the current state of the proof and resolve 2 problems related to the LeaveView step, in which we need to clear the contents of the Working Window (we have un-commented the steps related to View Change). This is problematic with the following 2 predicates we have in the Invariant - **SentPreparesMatchRecordedPrePrepareIfHostInSameView** and **EveryCommitIsSupportedByRecordedPrepares**
> 
> Thanks Hristo, just wondering - is it possible that these two predicates would need to be formulated differently, with view change in the picture? Maybe some of the previous claims don't hold anymore?

Yes, we need to formulate them in a way that knowledge will be preserved. The problem is as follows:
Once we introduced View Change into the model we have an action that clears the Replicas' Working Windows so that they can renegotiate the previous Views Working Window during the View Change process. 
We had invariants that were relying on the recorded information in the Working Window of the Replicas and those we need to restructure. Basically we must use the information in the Network model, since all messages sent from a host are always remembered there. This way the knowledge that justifies the sending of a Commit for a given View will be preserved.

teoparvanov reviewed (approved):
Looks great, Hristo! Establishing the basic invariants with view change enabled, as well as the byzantine replicas, is a major breakthrough!
